### PR TITLE
Prevent Hudi from accessing system tables of non-Hudi tables and other Hudi metadata refactor

### PR DESCRIPTION
**Pull request by findepi**

### CODE REVIEW

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiHiveViewsCompatibility.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import org.testng.annotations.Test;
+
+import static io.trino.tempto.assertions.QueryAssert.assertQueryFailure;
+import static io.trino.testing.TestingNames.randomNameSuffix;
+import static io.trino.tests.product.TestGroups.HUDI;
+import static io.trino.tests.product.TestGroups.PROFILE_SPECIFIC_TESTS;
+import static io.trino.tests.product.utils.QueryExecutors.onTrino;
+
+/**
+ * Tests interactions between Iceberg and Hive connectors, when one tries to read a view created by the other.
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiHiveViewsCompatibility.java:28]: `Iceberg and Hive` -> `Hudi and Hive`

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiMetadata.java [hunk 2]
```diff
@@ -149,14 +149,15 @@ private Optional<SystemTable> getRawSystemTable(SchemaTableName tableName, Conne
         if (!isHudiTable(session, tableOptional.get())) {
             return Optional.empty();
         }
-        switch (name.getTableType()) {
-            case DATA:
-                break;
-            case TIMELINE:
+        return switch (name.getTableType()) {
+            case DATA ->
+                // TODO remove DATA table type like https://github.com/trinodb/trino/commit/5dd017b4ebb668a63d42a45735d1caaaf566e179 for Iceberg and Delta
```
Comment (findinpath) [inline plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiMetadata.java:143]: Create a follow-up issue `good-first-issue` for this TODO task

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java [hunk 3]
```diff
@@ -1141,6 +1142,7 @@ public static boolean isHudiTable(Table table)
         String inputFormat = table.getStorage().getStorageFormat().getInputFormatNullable();
 
         return HUDI_PARQUET_INPUT_FORMAT.equals(inputFormat) ||
+                HUDI_MAPRED_INPUT_FORMAT.equals(inputFormat) ||
```
Comment (findinpath) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java:1145]: This is rather new to me. Separate commit pls.
How did you stumble upon it?

cc @codope did we miss this one while we agreed on the strategy for figuring out whether we're dealing with a Hudi table?

Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java]: it's a mistake actually

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiMetadata.java (hunk 2 above)
Comment (findepi) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiMetadata.java]: created https://github.com/trinodb/trino/issues/17973 but didn't make it gfi

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java [hunk 4]
```diff
@@ -1134,16 +1135,23 @@ public static boolean isIcebergTable(Map<String, String> tableParameters)
...
-        return HUDI_PARQUET_INPUT_FORMAT.equals(inputFormat) ||
+        if (HUDI_PARQUET_INPUT_FORMAT.equals(inputFormat) ||
                 HUDI_PARQUET_REALTIME_INPUT_FORMAT.equals(inputFormat) ||
                 HUDI_INPUT_FORMAT.equals(inputFormat) ||
-                HUDI_REALTIME_INPUT_FORMAT.equals(inputFormat);
+                HUDI_REALTIME_INPUT_FORMAT.equals(inputFormat)) {
+            return Optional.of(true);
+        }
+
+        if (PARQUET_MAPRED_INPUT_FORMAT.equals(inputFormat)) {
```
[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java:1151]: It's an attempt to close our split brain gap
https://github.com/trinodb/trino/blob/5295f3d5a66020f735e993911f743a5df860269f/plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java#L1137-L1147

https://github.com/trinodb/trino/blob/7da788222b4b34404abb76f81e75c8dd04a9f8d5/plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiUtil.java#L77-L91

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiHiveViewsCompatibility.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+        assertQueryFailure(() -> onTrino().executeQuery("SELECT * FROM hudi.default." + viewName))
+                .hasMessageMatching("Query failed \\(#\\w+\\):\\Q Not a Hudi table: default." + viewName);
+
+        onTrino().executeQuery("DROP VIEW hive.default." + viewName);
+        onTrino().executeQuery("DROP TABLE hive.default." + tableName);
+    }
+
+    @Test(groups = {HUDI, PROFILE_SPECIFIC_TESTS})
+    public void testHiveSelectFromHudiView()
```
Comment (ebyhr) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hudi/TestHudiHiveViewsCompatibility.java:52]: nit: The test name and the body don't match at a glance. Can we leave TODO comment that we want to add SELECT in Hive once Hudi connector supports creating views?

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiMetadata.java [hunk 6]
```diff
@@ -117,15 +116,25 @@ public HudiTableHandle getTableHandle(ConnectorSession session, SchemaTableName
...
         }
+
         if (table.getStorage().getOptionalLocation().isEmpty() || table.getStorage().getOptionalLocation().get().isEmpty()) {
             // No location or empty location cannot be a valid Hudi table
             return false;
         }
-        return HudiUtil.isHudiTable(fileSystemFactory.create(session), Location.of(table.getStorage().getLocation()));
+
+        Location location = Location.of(table.getStorage().getLocation());
+        return hudiMetadataExists(fileSystemFactory.create(session), location);
```
Comment (ebyhr) [inline plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiMetadata.java:137]: Is it possible to refactor like below? 
```java
    private boolean isHudiTable(ConnectorSession session, Table table)
    {
        Optional<Boolean> isHudiTable = HiveUtil.isHudiTable(table);
        if (isHudiTable.isPresent() && !isHudiTable.get()) {
            return false;
        }

        if (table.getStorage().getOptionalLocation().isEmpty() || table.getStorage().getOptionalLocation().get().isEmpty()) {
            // No location or empty location cannot be a valid Hudi table
            return false;
        }

        Location location = Location.of(table.getStorage().getLocation());
        if (!hudiMetadataExists(fileSystemFactory.create(session), location)) {
            throw new TrinoException(HUDI_BAD_DATA, "Location of table %s does not contain Hudi table metadata: %s".formatted(table.getSchemaTableName(), location));
        }
        return true;
    }
```

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiMetadata.java (hunk 6 above)
Comment (findepi) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiMetadata.java:137]: is the table is declared to use `HoodieParquetInputFormat`, it should be considered a hudi table, regardless whether we find `./.hoodie` inside.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java:1151]: @ebyhr good idea

@codope why Hudi tables are sometimes declared with `org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat` input format, same as normal Hive Parquet tables? How is one supposed to distinguish them?

File: plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiMetadata.java (hunk 6 above)
Comment (findepi) [reply plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiMetadata.java]: i rewrote this

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (findepi) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/HiveUtil.java]: based on https://github.com/trinodb/trino/blob/264ce58c9c23561beaa51ec8c27e8f9d83cf319f/plugin/trino-hudi/src/test/java/io/trino/plugin/hudi/testing/TpchHudiTablesInitializer.java#L172-L173 i concluded this is a hack to get test running. i fixed file metasotre instead.

**Pull request merged**